### PR TITLE
feat(tsi): Model Settings Implementation

### PR DESCRIPTION
**Pull request by bikamani**
This PR does not include e2e tests (follow-up PR)
Validated it works with sample code.

### CODE REVIEW

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+using Azure.Core;
+
+namespace Azure.Iot.TimeSeriesInsights.Customized
+{
+    /// <summary>
+    /// Request to update model settings. One of &quot;name&quot; or &quot;defaultTypeId&quot; must be set.
+    /// </summary>
+    [CodeGenModel("UpdateModelSettingsRequest")]
+    public partial class UpdateModelSettingsOptions
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs:12]: Renaming UpdateModelSettingsRequest to UpdateModelSettingsOptions

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the MIT License.
+
+using Azure.Core;
+
+namespace Azure.Iot.TimeSeriesInsights.Customized
+{
+    /// <summary>
+    /// Request to update model settings. One of &quot;name&quot; or &quot;defaultTypeId&quot; must be set.
+    /// </summary>
+    [CodeGenModel("UpdateModelSettingsRequest")]
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs:11]: changing visibility of this class to public 

If was internal initially it shows CS0051: inconsistent accessibility where the method UpdateModelSettings is public

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+
+namespace Azure.Iot.TimeSeriesInsights.Customized
+{
+    /// <summary>
+    /// Request to update model settings. One of &quot;name&quot; or &quot;defaultTypeId&quot; must be set.
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs:9]: Any recommendation for better description?

How about "Options to update model settings"?

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs [hunk 4]
```diff
@@ -26,12 +28,28 @@ public async Task RunSamplesAsync()
             try
             {
                 #region Snippet:TimeSeriesInsightsGetModelSettings
-                
+
                 // Get the model settings for the time series insights environment
-                Response<Models.ModelSettingsResponse> response = await client.GetAsync().ConfigureAwait(false);
-                Console.WriteLine($"Retrieved model {response.Value.ModelSettings.Name}.");
-                
+                Response<TimeSeriesModelSettings> currentSettings = await client.GetModelSettingsAsync().ConfigureAwait(false);
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs:33]: IIRC we leave out `.ConfigureAwait(false)` in our samples.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs [hunk 5]
```diff
@@ -26,12 +28,28 @@ public async Task RunSamplesAsync()
...
+                foreach (TimeSeriesIdProperty tsiId in currentSettings.Value.TimeSeriesIdProperties)
+                {
+                    Console.WriteLine($"Time series Id name: '{tsiId.Name}', Type: '{tsiId.Type}'.");
+                }
+
                 #endregion Snippet:TimeSeriesInsightsGetModelSettings
+
+                #region Snippet:TimeSeriesInsightsUpdateModelSettings
+
+                UpdateModelSettingsOptions options = new UpdateModelSettingsOptions();
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs:46]: var

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs [hunk 6]
```diff
@@ -26,12 +28,28 @@ public async Task RunSamplesAsync()
...
+                options.Name = "sampleModel";
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs:47]: use initializer list?

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs [hunk 7]
```diff
@@ -26,12 +28,28 @@ public async Task RunSamplesAsync()
...
+                Response<TimeSeriesModelSettings> updatedSettings = await client.UpdateModelSettingsAsync(options).ConfigureAwait(false);
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs:48]: Why not flatten the model and make them into params?

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 8]
```diff
@@ -86,6 +88,98 @@ public TimeSeriesInsightsClient(string environmentFqdn, TokenCredential credenti
             _timeSeriesTypesRestClient = new TimeSeriesTypesRestClient(_clientDiagnostics, _httpPipeline, environmentFqdn, versionString);
         }
 
+        /// <summary>
+        /// Gets model settings asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The Model Settings which includes model display name, Time Series ID properties and default type ID with the http response <see cref="Response{T}"/>.</returns>
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:95]: unless referring to a type by name and with a seealso reference, you don't need to capitalize them, so this should be `model settings`.

Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:95]: Casing Id

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 9]
```diff
@@ -86,6 +88,98 @@ public TimeSeriesInsightsClient(string environmentFqdn, TokenCredential credenti
...
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Updates model settings, either the model name or default type ID synchronously.
```
Comment (drwill-ms) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:160]: Id

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs (hunk 4 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples/TimeSeriesInsightsClientSample/TimeSeriesInsightsLifecycleSamples.cs:33]: Resolving as I keep adding changes locally.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+
+namespace Azure.Iot.TimeSeriesInsights.Customized
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/Customized/Models/UpdateModelSettingsOptions.cs:6]: namespace should match the namespace of the client code. So Azure.Iot.TimeSeriesInsights

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 11]
```diff
@@ -1,10 +1,12 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System;
 using System.Threading;
 using System.Threading.Tasks;
 using Azure.Core;
 using Azure.Core.Pipeline;
+using Azure.Iot.TimeSeriesInsights.Customized;
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:9]: Once you change the namespace of the customized class, you no longer need this.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 12]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
-        public virtual Task<Response<ModelSettingsResponse>> GetAsync(string clientSessionId = null, CancellationToken cancellationToken = default)
+        /// <returns>List of scopes for the specified endpoint.</returns>
+        internal static string[] GetAuthorizationScopes() => s_tsiDefaultScopes;
+
+        /// <summary>
+        /// Gets model settings asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The model settings which includes model display name, Time Series Id properties and default type Id with the http response <see cref="Response{T}"/>.</returns>
+        public virtual async Task<Response<TimeSeriesModelSettings>> GetModelSettingsAsync(CancellationToken cancellationToken = default)
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:109]: FYI, on my plate, I have a question that I will be asking the SDK board on whether we should be naming this GetTimeSeriesModelSettings, or leave as is. The same applies for the APIs I am working on for Instances. For example, GetInstances vs GetTimeSeriesInstances.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 13]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
-        /// </example>
-        public virtual Task<Response<ModelSettingsResponse>> GetAsync(string clientSessionId = null, CancellationToken cancellationToken = default)
+        /// <returns>List of scopes for the specified endpoint.</returns>
+        internal static string[] GetAuthorizationScopes() => s_tsiDefaultScopes;
+
+        /// <summary>
+        /// Gets model settings asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The model settings which includes model display name, Time Series Id properties and default type Id with the http response <see cref="Response{T}"/>.</returns>
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:108]: Response{TimeSeriesModelSettings}

Comment (barustum) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:108]: ..instead of Response{T}

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 14]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
+        public virtual async Task<Response<TimeSeriesModelSettings>> GetModelSettingsAsync(CancellationToken cancellationToken = default)
         {
-            return _modelSettingsRestClient.GetAsync(clientSessionId, cancellationToken);
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(GetModelSettings)}");
+            scope.Start();
+            try
+            {
+                Response<ModelSettingsResponse> modelSettings = await _modelSettingsRestClient.GetAsync(null, cancellationToken).ConfigureAwait(false);
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:115]: I recommend creating a task so we remember to fill in the clientSessionId instead of null. I added a clientSessionId for the client in my changes for Instances.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 15]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
+                TimeSeriesModelSettings timeSeriesModelSettings = modelSettings.Value.ModelSettings;
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:116]: nit and optional: no need to save modelSettings.Value.Modelsettings into a temporary variable, since it's only being used once in the next line.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 16]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
-        /// Response&lt;Models.ModelSettingsResponse&gt; response = await client.GetAsync().ConfigureAwait(false);
-        /// Console.WriteLine($&quot;Retrieved model {response.Value.ModelSettings.Name}.&quot;);
-        /// </code>
-        /// </example>
-        public virtual Task<Response<ModelSettingsResponse>> GetAsync(string clientSessionId = null, CancellationToken cancellationToken = default)
+        /// <returns>List of scopes for the specified endpoint.</returns>
+        internal static string[] GetAuthorizationScopes() => s_tsiDefaultScopes;
+
+        /// <summary>
+        /// Gets model settings asynchronously.
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:105]: Maybe reword to "Gets Time Series Insights model settings asynchronously.". I need to check with Eric on whether Time Series should be upper case or we can lower case it.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 17]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
         /// <summary>
-        /// Returns the model settings which includes model display name, Time Series ID properties and default type ID. Every Gen2 environment has a model that is automatically created.
+        /// Gets model settings synchronously.
         /// </summary>
-        /// <param name="clientSessionId"> Optional client session ID. Service records this value. Allows the service to trace a group of related operations across services, and allows the customer to contact support regarding a particular group of requests. </param>
-        /// <param name="cancellationToken"> The cancellation token to use. </param>
-        /// <returns>The deserialized application/json model settings digital twin and the http response <see cref="Response{ModelSettingsResponse}"/>.</returns>
-        public virtual Response<ModelSettingsResponse> Get(string clientSessionId = null, CancellationToken cancellationToken = default)
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The model settings which includes model display name, Time Series Id properties and default type Id with the http response <see cref="Response{T}"/>.</returns>
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:130]: Response{TimeSeriesModelSettings}

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 18]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
         }
 
         /// <summary>
-        /// Gets the scope for authentication/authorization policy.
+        /// Updates model settings, either the model name or default type Id asynchronously.
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:149]: Maybe reword to "Updates model settings model name or default type Id asynchronously."

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 19]
```diff
@@ -94,40 +96,101 @@ protected TimeSeriesInsightsClient()
...
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The updated model settings with the http response <see cref="Response{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException"> <paramref name="options"/> is null. </exception>
+        public virtual async Task<Response<TimeSeriesModelSettings>> UpdateModelSettingsAsync(UpdateModelSettingsOptions options, CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(UpdateModelSettings)}");
+            scope.Start();
+            try
+            {
+                Response<ModelSettingsResponse> modelSettings = await _modelSettingsRestClient.UpdateAsync(options, null, cancellationToken).ConfigureAwait(false);
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:161]: I would add a Argument.AssertNotNull(options) here.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 11 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:9]: makes sense. Thanks for catching this

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 12 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:109]: Ok does that mean we should hold on to this for now or check it in and change later?

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 14 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:115]: Sure, will take care of that.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 18 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:149]: How about Updates model name or default type Id async...

**Review by drwill-ms — APPROVED** [review]

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 14 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:115]: Added a to do for now.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 18 above)
Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:149]: wording to "Updates model name or default type Id on Time Series model settings"

**Review by barustum — APPROVED** [review]
